core/include/vecmem/containers/jagged_vector_data.hpp

```diff
@@ -0,0 +1,68 @@
+/*
+ * VecMem project, part of the ACTS project (R&D line)
+ *
+ * (c) 2021 CERN for the benefit of the ACTS project
+ *
+ * Mozilla Public License Version 2.0
+ */
+
+#pragma once
+
+#include "vecmem/memory/memory_resource.hpp"
+#include "vecmem/containers/details/jagged_vector_view.hpp"
+#include "vecmem/containers/device_vector.hpp"
+#include "vecmem/containers/vector.hpp"
+
+#include <cstddef>
+
+namespace vecmem {
+    template<typename T>
+    class jagged_vector_view;
+
+    /**
+     * @brief A data wrapper for jagged vectors.
+     *
+     * This class constructs the relevant administrative data from a vector of
+     * vectors, and is designed to be later turned into a @c jagged_vector_view
+     * object.
+     */
+    template<typename T>
+    class jagged_vector_data : public details::jagged_vector_view<T> {
```

Comment:
Should this class be in the `vecmem::` namespace? :thinking: I would've put it inside the `vecmem::details` namespace myself.

Mainly because I will still want to add a `vecmem::details::jagged_vector_buffer` type. One that will own all of the memory of the jagged vector that it describes. I absolutely agree though that it should not be part of this PR. (I still don't see some of the angles on that type...)

But since all the other "data types" are in the `vecmem::details` namespace, I would rather put this in there as well.

Reply:
But this data class (which is more like the view than the existing view class, which is potentially confusing) needs to be accessed by the user. Isn't that enough reason to have it under the `vecmem` namespace directly? I thought the `details` namespace was only for code the user shouldn't be using?

Reply:
I'm not sure myself... :frowning:

I would like to put the headers into a different directory than all of the "top level" host/device types. But you're absolutely right that the users will have to interact with these. For a while I thought that it could be possible for them to only interact with these types through `auto`. But no, that will not fly... :thinking:

I'm overall worried about all the classes making it difficult for a new user to get their bearing on what is available, and what everything does.

But I agree, shoving all of these types into `vecmem::details` is not a great solution either...

Reply:
We could always add more namespaces, but we'd have to think about it a little. Might I suggest we keep it like this for now, and we can always move them if we need to later?